### DISCUSSION TITLE
Convergence with manufactured solutions for convection-reaction problems

### DISCUSSION THREAD

**Discussion by rtaylo45** [idaholab/moose#17936] [Q&A General] [answered]
This is a long one and I am sorry. I am having trouble understanding the convergence rates for a test problem I am running. The problem is defined below 

![Screen Shot 2021-05-26 at 12 10 55 PM](https://user-images.githubusercontent.com/36215108/119694691-74542c80-be1b-11eb-9ba7-cec5c1362a5f.png)

This problem is very straight forward and using the method of characteristics has an analytic solution. I am trying to use the mms system to show verification with more complex reaction-convection problems but started with something simple. I have initially chosen this problem because it has an analytic solution that i can test the FEM solution against to show convergence. For this same problem I am also using the mms system to show convergence in space and time. 

Using the simple test function of `u = t^3 * x` I can get good temporal convergence rates for my mms runs and they agree with the temporal convergence rates for my analytical solution. Spatial convergence is what i am having trouble understanding. Using the analytical solution and a mms I conducted a spatial convergence study for three error metrics show below. 

![progression_02-01_spatial_refinement_analytical](https://user-images.githubusercontent.com/36215108/119697807-98653d00-be1e-11eb-8a20-f154b3086214.png)

![progression_02-01_spatial_refinement](https://user-images.githubusercontent.com/36215108/119701104-ffd0bc00-be21-11eb-8928-ab792214d56c.png)

For convection I am using LevelSetAdvection and LevelSetAdvectionSUPG. Looking at the LevelSet module, they have a mms test case where they show second order convergence on the default Lagrange mesh with a similar PDE but with out the decay source term. I am guessing that because i am doing a more complex PDE it will not show a higher rate of convergence against either mms or the true analytical solution? Another question I have is why the convergence rates using mms and analytical don't match up more? Now they do both follow the same general trend however I am very new to using mms and just want to make sure that what I am seeing is ok. The analytical solution is more simple than my assumed mms, not sure how much that changes the convergence rates.  I don't think the LevelSet module us the AD method for the jacobian and not sure if i should be overwriting the class and add new derivatives to the jacobian calculation. 

In summary I have two questions. Should this problem be converging at a fast rate? Why is the convergence rate for mms slightly worse than testing convergence with the analytical solution?

I have adjusted the time step sizes to make sure that they are low enough for this problem. Attached are the two input files saved as txt files because of github. 

[progression_02-01_mms_spatial.txt](https://github.com/idaholab/moose/files/6548454/progression_02-01_mms_spatial.txt)
[progression_02-01.txt](https://github.com/idaholab/moose/files/6548455/progression_02-01.txt)

**Comment by jiangwen84** [comment, marked as answer]
@rtaylo45 Nice benchmark problems.

If you want to use SUPG stabilization, you need to add all SUPG terms, i.e., TimeDerivative, Forcing Function and Advection Function. In your current input file, you only have Advection SUPG term. See `modules/level_set/test/tests/verification/1d_level_set_supg_mms/1d_level_set_supg_mms.i` for example. 

For this problem, you probably do not need SUPG stabilization. Remove it and give it a shot.

**Reply by rtaylo45** [reply]
If i were to still use the SUPG stabilization would i need to write a kernel for the decay term in the equation?

**Reply by jiangwen84** [reply]
Looks like you can reuse `LevelSetForcingFunctionSUPG` for the decay term.

**Reply by rtaylo45** [reply]
So based on the `LevelSetForcingFunctionSUPG` source file it is based on the `ADKernelGrad` class. This is confusing to me because based on the formalization in the source documents it seems like they should be based on the `ADKernelValue` class because it shouldn't use the grad of the nonlinear variable `u` but the value of `u`. It seems like it might have something to do with the return value of `ADRealVectorValue`. I wondering if it might be useful to rewrite these `LevelSetSUGP` classes using the `ADKernelStabilizedTempl` base class which has an virtual function `computeQpStabilization()` method which has a return type of `ADRealVectorValue`.

**Reply by jiangwen84** [reply]
I think my previous comment is not correct. The `LevelSetForcingFunctionSUPG` only works for a source term. For your case, you probably need to implement a new SUPG kernel for the decay term. 

The `LevelSetForcingFunctionSUPG` implements the weak form given as -tau * velocity * grad_test * f. If we use `ADKernelGrad`, then we only need to provide  (-tau * velocity * f)  to `precomputeQpResidual`.

It seems to be possible to write `LevelSetSUPG` using `ADKernelStablized`, although I do not see any obvious advantage there. `ADKernelStablized` was used for shallow water equation before. Maybe @aeslaughter would know this better than me.

**Reply by rtaylo45** [reply]
Ok I think I see what i was missing, the `grad_test` part was coming from the stabilization term.

**Reply by aeslaughter** [reply]
@jiangwen84 knows more than I do at this point, it has been many moons since I did that work and it would take me a bit to recall what was doing. However, I remember updating the level set module when I did the work. So, I am assuming that those are implemented with the desired base classes.
